src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs

```diff
@@ -141,9 +144,11 @@ public HttpHeaderValueCollection<WarningHeaderValue> Warning
 
         #endregion
 
-        internal HttpResponseHeaders()
-            : base(HttpHeaderType.General | HttpHeaderType.Response | HttpHeaderType.Custom, HttpHeaderType.Request)
+        internal HttpResponseHeaders(bool containsTrailingHeaders = false)
+            : base(containsTrailingHeaders ? HttpHeaderType.All ^ HttpHeaderType.Request : HttpHeaderType.General | HttpHeaderType.Response | HttpHeaderType.Custom,
```

Comment:
Enabling `HttpHeaderType.Content` here will cause `TryAddWithoutValidation("Content-Length", "123")` and `GetValues("Content-Length")` to succeed where they will currently fail for the trailing headers collection.

Reply:
Suggestion here might work to resolve this: https://github.com/dotnet/corefx/pull/42229/files#r340850684

Reply:
This is caused by the issue with an incomplete disallowed header list in HttpConnection I reported yesterday. Currently, I just re-use that list of prohibited headers, however it misses some headers mentioned in the RFC (e.g. Proxy-Authorization). Content-Length is just another example.
It seems introducing new HttpHeaderType.NonTrailing is a good option. However, I have one concern. Currently, HttpHeaderType enum have mutually exclusive fields with only one exception of All. If I add NonTrailing flag, it will be set in combination with others (e.g. Request and NonTrailing), and it won't be clear how All should behave in this case. It seems NonTrailing represents a different concept because it doesn't define a header type (i.e. splits the set of all existing header into non-intersecting subsets ), but instead specify a permission for the given header instance.
Don't you think it could cause problems in the future? How will we treat All value? As "any header type" & "non in trailer" or as "any header" & "in trailer"?

Reply:
`All` is only used in a single place right now, so I'm not really worried about it.

However there are places in code that do an explicit e.g. `if(headerType == HttpHeaderType.Response)` that would need to be changed to a `HasFlag`. I don't see this as a problem.

Reply:
Even though ```All``` is use only in one place now,  the introduction of NonTrailing flag still breaks its abstraction. Let alone the possibility that it can be used wider later.

Reply:
@stephentoub What do you think about the above proposal to introduce NonTrailing flag and the concerns around it?

Reply:
This way seems no-cost from a perf perspective, but I'm not tied to it if we can think of a better way.

I don't understand what you mean by abstraction -- do you mean you wouldn't want to see the enum used as flags? It already is, but perhaps changing the name to `HttpHeaderFlags` would make things clearer.

Reply:
I think adding `HttpHeaderType.NonTrailing` is fine as it makes things very clear in the code. `All` isn't used much. And this is all internal code anyways. So, the priority should be performance and code maintenance.

Reply:
@scalablecory An abstraction is a generalized description of some entities in the given set which defines a common set of rules and statements valid for all entities in the set. In the given case, ```All``` means "all headers which can appear in HTTP messages". If we introduce the ```NonTrailing``` flag to the same ```HttpHeaderType```, it breaks this statement because we have to explicitly set either 0 or 1 value for the respective bit in ```All``` definition which will lead to one of the following mutually exclusive definitions:

1. "all headers which can appear in HTTP messages and can be present in the trailer"
2. "all headers which can appear in HTTP messages and cannot be present in the trailer"

Reply:
That said, it seems an introduction of ```NonTrailing``` completely eliminates the necessity to have HashSet, so I change the implementation. However, it has another (more practical) disadvantage, now it's mandatory to use bit ```&``` for any header type comparison. Please, check the current PR version and let me know your opinion if the perf improvement really outweighs that inconvenience.

Reply:
@alnikola It looks like your latest commit has chosen (2) ""all headers which can appear in HTTP messages and cannot be present in the trailer""? Is that right?  If so, perhaps it makes sense to rename the `All` bit to something else like `AllNonTrailing`?

Reply:
That's correct, however it's only because I had no other choice but to assign either 0 or 1. I'm not sure what it's the best option. Overall, do you think that getting rid of HashSet really justifies that ambiguity and the requirement to always compare header types through the bitwise ```&```?

Reply:
Since `All` is only used in one place, is it reasonable to get rid of it entirely?

Reply:
What's the concern about using `&` to check for header type?

Reply:
>Please, check the current PR version and let me know your opinion if the perf improvement really outweighs that inconvenience.

Since this isn't a public API, the potential usability concerns for the "inconvenience" aren't very strong. So, the perf improvment does outweigh other concerns. I think we should either 1) Remove the `All` if not really needed, or 2) rename the `All` to something more descriptive. And there should be more Assert's added in the code to help with future code maintenance/refactoring.

Reply:
It seems there is a misunderstanding. My main concern is not about ```All```, but about the fact that I had to change all header type comparison expressions by introducing bit-wise ```&``` operator because now it's impossible to directly compare with via a simple == expression. This approach is really fragile because any new person unaware about this approach will do it wrong.
In example,

- Before ```descriptor.HeaderType == HttpHeaderType.Content```
- **Now** ```(descriptor.HeaderType & HttpHeaderType.Content) == HttpHeaderType.Content```

Please, check all the changes made in [my latest commit](https://github.com/dotnet/corefx/pull/42229/commits/2a762ea4f00c9cf264e810e7a9124d833af7db6a).

Reply:
>This approach is really fragile because any new person unaware about this approach will do it wrong.

I don't think it's fragile. I do agree that it is a change from how it worked before. But basically these are now working as proper bit flags which do require '&' operators to verify that the bit is set.

Perhaps you could write some helper methods (which marked correctly with attributes would be inline-able by the compiler).

To help a new person get it right, just add some Debug.Assert's in the right places. Those will help make sure that future coding changes adhere to the new pattern.

Reply:
Thanks.